src/client/ContentManagement.js

```diff
@@ -621,55 +619,59 @@ exports.CreateContentObject = async function({libraryId, objectId, options={}})
  */
-exports.CopyContentObject = async function({libraryId, originalVersionHash, options={}, originalObject=null}) {
+exports.CopyContentObject = async function({libraryId, originalVersionHash, options={}}) {
   ValidateLibrary(libraryId);
   ValidateVersion(originalVersionHash);
 
   options.copy_from = originalVersionHash;
 
-  if(!originalObject) {
-    return await this.CreateAndFinalizeContentObject({libraryId, options});
-  }
-
-  ValidateObject(originalObject.id);
+  const {objectId, writeToken} = await this.CreateContentObject({libraryId, options});
+  const originalObjectId = this.utils.DecodeVersionHash(originalVersionHash).objectId;
+  const metadata = await this.ContentObjectMetadata({versionHash: originalVersionHash});
+  const permission = await this.Permission({objectId: originalObjectId});
 
-  const callback = async ({objectId, writeToken}) => {
-    const kmsCapKey = Object.keys(originalObject).find(key => key.startsWith("eluv.caps.ikms"));
-    const userCapKey = Object.keys(originalObject).find(key => key.startsWith("eluv.caps.iusr"));
+  // User CAP
+  const hasUserCap = Object.keys(metadata).some(key => key.startsWith("eluv.caps.iusr"));
+  const userCapKey = `eluv.caps.iusr${this.utils.AddressToHash(this.signer.address)}`;
 
-    if(userCapKey) {
-      const userConkKey = await this.Crypto.DecryptCap(originalObject[userCapKey], this.signer.signingKey.privateKey);
-      userConkKey.qid = objectId;
+  if(hasUserCap) {
+    if(!metadata[userCapKey]) {
+      throw Error("Unable to find a user cap for the current user");
+    }
 
-      await this.DeleteMetadata({
-        libraryId,
-        objectId,
-        writeToken,
-        metadataSubtree: userCapKey
-      });
+    const isOwner = this.utils.EqualAddress(this.signer.address, await this.ContentObjectOwner({objectId: originalObjectId}));
 
-      this.ReplaceMetadata({
-        libraryId,
-        objectId,
-        writeToken,
-        metadataSubtree: userCapKey,
-        metadata: await this.Crypto.EncryptConk(userConkKey, this.signer.signingKey.publicKey)
-      });
+    if(!isOwner) {
+      throw Error(`Current user is not owner of object ${metadata}`);
     }
 
-    if(kmsCapKey) {
-      await this.DeleteMetadata({
-        libraryId,
-        objectId,
-        writeToken,
-        metadataSubtree: kmsCapKey
-      });
+    const userConkKey = await this.Crypto.DecryptCap(metadata[userCapKey], this.signer.signingKey.privateKey);
+    userConkKey.qid = objectId;
 
-      await this.SetPermission({objectId, permission: originalObject.permission, writeToken});
-    }
-  };
+    this.ReplaceMetadata({
+      libraryId,
+      objectId,
+      writeToken,
+      metadataSubtree: userCapKey,
+      metadata: await this.Crypto.EncryptConk(userConkKey, this.signer.signingKey.publicKey)
+    });
+  }
+
+  // KMS CAP
+  const kmsCapKey = Object.keys(metadata).find(key => key.startsWith("eluv.caps.ikms"));
```

Comment:
This should probably delete all KMS caps it comes across. You can do something like this:

```
await Promise.all(
  Object.keys(metadata)
  .filter(key => key.startsWith("eluv.caps.ikms"))
  .map(async kmsCapKey =>  await this.DeleteMetadata({
      libraryId,
      objectId,
      writeToken,
      metadataSubtree: kmsCapKey
    })
)
```